test/libsolidity/util/SoltestTypes.h

```diff
@@ -0,0 +1,256 @@
+*/
+
+#pragma once
+
+#include <libdevcore/CommonData.h>
+#include <libsolidity/ast/Types.h>
+
+namespace dev
+{
+namespace solidity
+{
+namespace test
+{
+
+/**
+ * All soltest tokens.
+ */
+#define SOLT_TOKEN_LIST(T, K)      \
+	T(Unknown, "unknown", 0)       \
+	T(Invalid, "invalid", 0)       \
+	T(EOS, "EOS", 0)               \
+	T(Whitespace, "_", 0)          \
+	/* punctuations */             \
+	T(LParen, "(", 0)              \
+	T(RParen, ")", 0)              \
+	T(LBrack, "[", 0)              \
+	T(RBrack, "]", 0)              \
+	T(LBrace, "{", 0)              \
+	T(RBrace, "}", 0)              \
+	T(Sub,    "-", 0)              \
+	T(Colon,  ":", 0)              \
+	T(Comma,  ",", 0)              \
+	T(Period, ".", 0)              \
+	T(Arrow, "->", 0)              \
+	T(Newline, "//", 0)            \
+	/* Literals & identifier */    \
+	T(Comment, "#", 0)             \
+	T(Number, "number", 0)         \
+	T(HexNumber, "hex_number", 0)  \
+	T(String, "string", 0)         \
+	T(Identifier, "identifier", 0) \
+	/* type keywords */            \
+	K(Ether, "ether", 0)           \
+	K(Hex, "hex", 0)               \
+	K(Boolean, "boolean", 0)       \
+	/* special keywords */         \
+	K(Left, "left", 0)             \
+	K(Right, "right", 0)           \
+	K(Failure, "FAILURE", 0)       \
+
+namespace soltest
+{
+	enum class Token : unsigned int {
+	#define T(name, string, precedence) name,
+		SOLT_TOKEN_LIST(T, T)
+		NUM_TOKENS
+	#undef T
+	};
+
+	/// Prints a friendly string representation of \param _token.
+	inline std::string formatToken(Token _token)
+	{
+		switch (_token)
+		{
+	#define T(name, string, precedence) case Token::name: return string;
+			SOLT_TOKEN_LIST(T, T)
+	#undef T
+			default: // Token::NUM_TOKENS:
+				return "";
+		}
+	}
+}
+
+/**
+ * The purpose of the ABI type is the storage of type information
+ * retrieved while parsing a test. This information is used
+ * for the conversion of human-readable function arguments and
+ * return values to `bytes` and vice-versa.
+ * Defaults to None, a 0-byte representation. 0-bytes
+ * can also be interpreted as Failure, which means
+ * either a REVERT or another EVM failure.
+ */
+struct ABIType
+{
+	enum Type
+	{
+		None,
+		Failure,
+		Boolean,
+		UnsignedDec,
+		SignedDec,
+		Hex,
+		HexString,
+		String
+	};
+	enum Align
+	{
+		AlignLeft,
+		AlignRight,
+		AlignNone,
+	};
+	Type type = ABIType::None;
+	Align align = ABIType::AlignRight;
+	size_t size = 0;
+	bool alignDeclared = false;
+};
+
+/**
+ * Helper that can hold format information retrieved
+ * while scanning through a parameter list in soltest.
+ */
+struct FormatInfo
+{
+	bool newline = false;
+};
+
+/**
+ * Parameter abstraction used for the encoding and decoding of
+ * function parameter and expectation / return value lists.
+ * A parameter list is usually a comma-separated list of literals.
+ * It should not be possible to call create a parameter holding
```

Comment:
"call create"?

Reply:
Fixed in https://github.com/ethereum/solidity/pull/7052/.